Add `user` option to `Redis` adapter to authenticate with a specific user

vcg-development opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Assuming the current release is 1.5.0, the next patch release is 1.5.1, the next minor is 1.6.0 and the next major is 2.0.0, The Current release branch will be `1.5.x`, the next minor branch will be `1.6.x`, and the next major branch will be `2.0.x`

Pick the target branch based on the following criteria:
  * Documentation improvement: Current release branch 1.5.x
  * Bugfix: Current release branch 1.5.x
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: Next minor 1.6.x
  * New feature, or refactor of existing code: Next minor 1.6.x
  * Backwards incompatible features and refactoring: Next major 2.0.x

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

Description
This PR is adding the possibility to authenticate with Redis with a custom username. The ACL feature is available in Redis since Version 6.
The username can be passed as an resource option.

<!--

Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE CURRENT RELEASE BRANCH

- Are you adding documentation?
  - TARGET THE CURRENT RELEASE BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE NEXT MINOR BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE CURRENT RELEASE BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE NEXT MINOR BRANCH OR THE NEXT MAJOR IF BC WILL BE BROKEN

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE NEXT MINOR BRANCH OR THE NEXT MAJOR IF BC WILL BE BROKEN
-->

cbraunVBM committed:
Added option to set redis user name in redis options

Signed-off-by: Christian Braun <christian.braun@vogel.de>

cbraunVBM committed:
whitespace removed

boesing commented on src/RedisOptions.php:
```diff
@@ -261,4 +260,26 @@ public function getPassword()
+     * @param string $user ACL User
+     * @return RedisOptions Provides a fluent interface
+     */
+    public function setUser($user)
```
Since this is a new method, we can add argument type here.
```suggestion
    public function setUser(string $user): void
```

I'd also remove `return $this;` so we can introduce `: void` return type.

vcg-development replied:
I added the argument type, but kept the RedisOptions return type to be compatible with the other setters and getters

boesing replied:
Fine for me, will delete them in v4 anyways.

boesing commented on src/RedisOptions.php:
```diff
@@ -261,4 +260,26 @@ public function getPassword()
+     *
+     * @return string
+     */
+    public function getUser()
```
Is always a `string` returned, even if nothing was set beforehand?
If so, please add native return type.
If `null` is returned if user was not previously configured via `setUser`, the return value should reflect all possible return values.

boesing commented on src/RedisResourceManager.php:
```diff
@@ -123,6 +123,22 @@ public function getPassword($id)
+     * @param string $id
+     * @return string
+     */
+    public function getUser($id)
```
Since this is a new method, we can add argument type here.
```suggestion
    public function getUser(string $id): string
```

boesing commented on src/RedisResourceManager.php:
```diff
@@ -279,6 +295,36 @@ protected function extractPassword($resource, $serverUri)
+     */
+    protected function extractUser($resource, $serverUri)
+    {
+        if (! empty($resource['user'])) {
```
I wonder if we can improve the codebase a bit. Since we are expecting `resource` to be an array, we should require `array` in the method signature.
The method is called explicitly and thus we can already do an `is_array()` check beforehand and/or add a proper array shape annotation to the `$resource` storage which reflects the actual type, such as: `array{user?:string,...}`

I can prepare a patch at some point to show what I mean but in case you know what I am talking about, feel free to improve some types on your own.

boesing commented on src/RedisResourceManager.php:
```diff
@@ -279,6 +295,36 @@ protected function extractPassword($resource, $serverUri)
+        // parse server from URI host{:?port}
+        $server = trim($serverUri);
+
+        if (strpos($server, '/') === 0) {
```
you can use `str_starts_with` which is also more readable.
Just a nitpick, not a strong requirement tho.

boesing commented on src/RedisResourceManager.php:
```diff
@@ -644,7 +699,7 @@ public function setServer($id, $server)
      * @param string $id
      * @param string $password
-     * @return RedisResource
+     * @return RedisResourceManager
```
good catch!

boesing commented on src/RedisResourceManager.php:
```diff
@@ -713,4 +768,25 @@ private function getRedisInfo(RedisResource $redis): array
+     * @param string $user
+     * @return RedisResourceManager
+     */
+    public function setUser($id, $user)
```
Since this is a new method, we can add argument type here.
```suggestion
    public function setUser(string $id, string $user): void
```

I'd also remove `return $this;` so we can introduce `: void` return type.

boesing commented on test/unit/RedisResourceManagerTest.php:
```diff
@@ -71,6 +72,27 @@ public function testSetServerWithPasswordInParameters(): void
+        $this->assertEquals('dummyuser', $this->resourceManager->getUser($dummyResId2));
+    }
+
+    public function testSetServerWithPasswordInParametersAndNoUser(): void
```
can you add a test to verify that extracting the user from a URI works as expected?

You introduced a way to extract the username from `redis://username:password@testhost:1234` but I do not see a test for that.

boesing commented on test/unit/RedisResourceManagerTest.php:
```diff
@@ -287,6 +309,36 @@ public function testWillCatchAuthDuringConnectException(): void
+            ->expects(self::atLeastOnce())
+            ->method('auth')
+            ->with(['dummyuser', 'secret'])
+            ->willThrowException(new RedisException('test'));
```
why would that method call throw an exception?

If you have `expects` defined so that it is verified that the method is called for sure, that is already an assertion.
By providing `with`, that would explicitly check that when the method is invoked, all arguments are matching the expectations we passed here. 

Is the test failing when you remove that exception?

boesing reviewed (changes requested):
Sorry for the late feedback, I didn't felt like having a look into it.
However, thanks for this contribution. I wonder if it is also available for `RedisCluster` but I do not have a problem if that is a `Redis`-only feature for now, can be implemented for `RedisCluster` once requested.

I do have some nitpicks and some change requests.

- especially the native return types and argument types should be added
- we should have a test which verifies that the user can be extracted from the server URI

I think its time to have a proper resource array shape declared in the `RedisResourceManager` so that we do not have that weird `array` with "whatever" array keys in it. Once I find some time, I might implement that myself and push it to this patch - if you think you want to that on your own, I'd be happy to review that change again.

We recently introduced `SSLContextArrayShape` in `SslContext`. You can take that as an example on how to declare array shapes if you do not have an idea on how that might work.

Once we defined the `RedisResourceArrayShape`, we can use that for `RedisResourceManager::$resources`. That will most probably remove a bunch of baseline entries and errors within the `RedisResourceManager` so that would be a huge improvement.

boesing commented on src/RedisResourceManager.php:
```diff
@@ -315,7 +361,9 @@ protected function connect(array &$resource)
             $resource['initialized'] = true;
-            if ($resource['password']) {
+            if ($resource['user'] && $resource['password']) {
+                $redis->auth([$resource['user'], $resource['password']]);
```
I checked the phpredis [`redis.stub.php`](https://github.com/phpredis/phpredis/blob/62cf943fecc5182c6329b332df43cf28012cef55/redis.stub.php#L586C1-L587C1) and if I understand it correct, the value has to be an array:
```php
$redis->auth(['username', 'password']);
```

So I guess we have to adapt here and also fix the unit test.

cbraunVBM committed:
implemented changes from pull request

Signed-off-by: Christian Braun <christian.braun@vogel.de>

cbraunVBM committed:
added ResourceArrayShape

Signed-off-by: Christian Braun <christian.braun@vogel.de>

cbraunVBM committed:
added missing test

Signed-off-by: Christian Braun <christian.braun@vogel.de>

cbraunVBM committed:
removed unnecessary test

Signed-off-by: Christian Braun <christian.braun@vogel.de>

cbraunVBM committed:
Changed redis auth to array

Signed-off-by: Christian Braun <christian.braun@vogel.de>

vcg-development commented:
Thank you very much for the detailed review. I changed the code according to your comments.

I also added the suggested ResourceArrayShape. I`d be happy, if have another look at my changes.

boesing commented on src/RedisResourceManager.php:
```diff
@@ -22,17 +22,33 @@
+/**
+ * @psalm-type ResourceArrayShape =array{
+ *     resource?: RedisResource,
+ *     database?: non-empty-string,
```
aren't databases integers?
I guess `non-negative-int` would be the correct type, but since we are only juggling with `int` so far, I'd be fine with at least having this type `int`.

boesing commented on src/RedisResourceManager.php:
```diff
@@ -22,17 +22,33 @@
-     * Registered resources
+     *  Registered resources
+     *
+     * @var array<ResourceArrayShape> $resources
```
```suggestion
     * @var array<string, ResourceArrayShape> $resources
```
the key here should be of the same type of `resource` which seems to be string according to the other methods in this class

boesing commented on src/RedisResourceManager.php:
```diff
@@ -22,17 +22,33 @@
+ *     resource?: RedisResource,
+ *     database?: non-empty-string,
+ *     lib_options?: array,
+ *     version?: int|string,
```
hm, I am unsure if I've ever seen `int` returned from redis itself but I also never tested that.
Where did you get that type from?

vcg-development replied:
defaults initialized version with 0, i changed this to '' now

boesing commented on src/RedisResourceManager.php:
```diff
@@ -22,17 +22,33 @@
  * This is a resource manager for redis
  */
+/**
+ * @psalm-type ResourceArrayShape =array{
```
I am almost sure that at least a bunch of keys are *always* available as there is that `$defaults` array shape declared in `RedisResourceManager#setResource`.
So I'd say the keys in that shape (which are present in `$defaults` are always available - so we can remove the `?` which is used to outline that keys might be missing).

```
  $defaults = [
            'persistent_id' => '',
            'lib_options'   => [],
            'server'        => [],
            'password'      => '',
            'database'      => 0,
            'resource'      => null,
            'initialized'   => false,
            'version'       => 0,
        ];
```

boesing commented on src/RedisResourceManager.php:
```diff
@@ -22,17 +22,33 @@
+ *     version?: int|string,
+ *     password?: non-empty-string,
+ *     persistent_id?: non-empty-string,
+ *     server?: non-empty-string,
```
`server` is actually an array which is getting normalized to:

```php
$server = [
    'host'    => $host,
    'port'    => $port,
    'timeout' => $timeout,
];
```

within `normalizeServer`. 

```suggestion
 *     server: array{host: string, port: int, timeout: int},
```

boesing commented on src/RedisResourceManager.php:
```diff
@@ -315,8 +371,10 @@ protected function connect(array &$resource)
+            if ($resource['user'] && $resource['password']) {
+                $redis->auth([$resource['user'], $resource['password']]);
+            } elseif ($resource['password']) {
+                $redis->auth([$resource['password']]);
```
we could've kept the original implementation as `string` still works for `password`-only.

boesing replied:
_(not a task, just FYI)_

boesing reviewed (changes requested):
Nice, thanks for updating and providing proper type-hints for the new methods!
I put some suggestions and comments to your latest changes regarding the array shape. If you apply these changes, I guess psalm will stop reporting some errors.

You might want to have a look on these psalm issues as they might tell you why some types may probably not correct 👍🏼 

If you need help, just let me know.

cbraunVBM committed:
Corrected Types in ResourceArrayShape and linked function return types

Signed-off-by: Christian Braun <christian.braun@vogel.de>

cbraunVBM committed:
fixed test

Signed-off-by: Christian Braun <christian.braun@vogel.de>

vcg-development commented:
Thank you again for having a deep look at my commits again. The types declared in the array shape and related return types should be correct now. 
I thing the failing psalm checks come from the array_merge in line 416. At the moment I am not sure how this could be fixed, so that they will not fail anymore.

Also I think the Assertment in Line 177 is not necessary anymore, because the instance is already checked in line 169.

boesing committed:
qa: update `psalm-baseline.xml` and apply some changes for psalm

Due to the way how this class worked before, we make an assumption that a provided `iterable` will contain at least those keys we are supporting.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
Yes you are right, one of the issues were due to the `array_merge` (psalm has issues with that while `array_replace` does work. In this case, `array_replace` should suffice and thus I changed the underlying code.

I also removed the fixed issues with psalm from the baseline by updating the baseline using `psalm --update-baseline`. That will automatically remove all fixed issues which were fixed by this PR.

I will have a closer look on the integration tests, it seems that the password test is actually introducing problems when used on a server without a configured password.

boesing committed:
bugfix: always store `$resource` in internal property cache

Due to refactoring, the property reference was removed but the resource was only re-written to the propery cache when the version was not available. This is fixed with this commit.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add early return

To have the same functionality as in other methods (which provide fluent interface), we should use early return to prevent unsetting `initialized` logic.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: apply coding standard

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: only mark connection as initialized when authentication and database selection was executed

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: restructure redis related tests to prevent side-effects

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove obsolete `tearDown` implementation as we do not unset `storage` property anymore

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: make `RedisOptions` tests pass

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
Okay, that was a tough one. I was wondering why stuff started to went wrong after my changes.
I found out that in one of my changes, I removed the property reference which had side-effects in some tests.

Another good reason on why having these property references removed from our code.
I've created https://github.com/laminas/laminas-cache-storage-adapter-redis/issues/93 so that we can tackle further refactorings in v3.0.0.

boesing commented:
Thanks @vcg-development and @cbraunVBM!
I will create an issue in laminas-cache so that we can document the new option as that has to be done there.
